map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,172 @@
+#include "map/extrapolation/extrapolator.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/logging.hpp"
+
+#include <chrono>
+#include <memory>
+
+namespace
+{
+uint64_t constexpr kMaxExtrapolationTimeMs = 1000;
+uint64_t constexpr kExtrapolationPeriodMs = 200;
+double constexpr kMaxExtrapolationSpeedMPS = 120.0;
+
+double Delta(double param1, double param2, uint64_t timeBetweenPointsMs, uint64_t timeAfterParam2Ms)
+{
+  return (param2 - param1) * timeAfterParam2Ms / timeBetweenPointsMs;
+}
+
+class LinearExtrapolator
+{
+public:
+  LinearExtrapolator(uint64_t timeBetweenMs, uint64_t timeAfterMs)
+    : m_timeBetweenMs(timeBetweenMs), m_timeAfterMs(timeAfterMs)
+  {
+  }
+
+  double Extrapolate(double x1, double x2)
+  {
+    return x2 + (x2 - x1) / m_timeBetweenMs * m_timeAfterMs;
+  }
+
+private:
+  uint64_t m_timeBetweenMs;
+  uint64_t m_timeAfterMs;
+};
+}  // namespace
+
+namespace extrapolation
+{
+using namespace std;
+
+location::GpsInfo LinearExtrapolation(location::GpsInfo const & gpsInfo1,
+                                      location::GpsInfo const & gpsInfo2,
+                                      uint64_t timeAfterPoint2Ms)
+{
+  auto const timeBetweenPointsMs =
+      static_cast<uint64_t>((gpsInfo2.m_timestamp - gpsInfo1.m_timestamp) * 1000);
```

Comment:
Please add a CHECK that the second timestamp is not less than the first.

Reply:
`timeBetweenPointsMs` shouldn't be zero. So the second timestamp should be more than the first one.

Reply:
I added `CHECK_LESS(gpsInfo1.m_timestamp, gpsInfo2.m_timestamp, ());`